### PR TITLE
feat: trigger scheduled tasks according to controllable clock

### PR DESCRIPTION
**Pull request by lenaschoenburg**
The implementation is admittedly a bit crude but it works and I hope that it's good enough.
The idea is that `ProcessingScheduleServiceImpl` does not "forward" scheduled tasks to the underlying actor scheduler where we don't have control over the clock, but that it schedules them internally and only has the actor scheduler periodically call to check if any timers are now expired. That way, we have control over the clock we use to decide if a scheduled tasks is expired or not.

Right now the service is called every second. If a task is scheduled for less than half of that, we schedule an additional check a bit earlier, right when the task is supposed to expire. This is purely an optimization, mostly done so that we don't have to adjust a lot of tests that assume that something scheduled with zero delay will trigger more or less immediately.

Relates to https://github.com/camunda/camunda/issues/21051

### CODE REVIEW

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:43]: :question: Is this unbounded? If it is, can we either:

1. Bound it. This means somehow handling issues when it _is_ bounded...
2. Have some visibility on its size, e.g. a metric or something. This could help speed up memory issue diagnosis in the future.

Comment (npepinpe) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:43]: If we think memory could be an issue, we could try and compare quickly a full `DeadlineTimerWheel` with 1 million entries memory size vs this priority queue. There are some easy utilities that tell you the size of an object (e.g. https://github.com/openjdk/jol/blob/master/README.md)

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:213]: :heart:

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:214]: :question: Would making it a record reduce its memory footprint in the priority queue? I know the actual memory difference is generally small, but may be more at scale - and I believe (I'd have to look up which Java quark went into it) that the GC handles records more efficiently, so it might be a better choice overall for storing potentially many of them in an unbounded collection.

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (npepinpe) [inline zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:116]: :question: Will you make this configurable in a follow up PR as discussed?

**Review by npepinpe — APPROVED** [review]
:rocket: Please address comments, but nothing that requires a second review I think :)

My main concern as usual is the unbounded queue :sweat_smile:

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:43]: Bounding it would be difficult because nothing is set up to handle failure cases. Just as an example, what would it even mean if a `DueDateChecker` cannot schedule next execution?

For size comparison:
Before this PR, every scheduled timer would end up as an entry in the `DeadlineTimerWheel` (roughly two longs + inevitable overhead) _and_ a `TimerSubscription` on the actor task. Both implementations of `TimerSubscription` require 46 bytes according to JOL.

After this PR, every scheduled timer is a `ScheduledTaskImpl` in the priority queue, each taking 24 bytes according to JOL.

Overall this is a reduction in memory usage, with the same danger of unbounded growth as before.

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:214]: I can't find a source for this. The only benefit I'm aware of is that the JVM trusts that record fields are immutable (unlike regular object fields) and can apply more optimizations that way.

I've deliberately chosen not to use a record because it's not just holding state but allows cancellation too.


I'm happy to change it to a record though if you can find a source for memory/gc benefits :)

**Comment by lenaschoenburg** [comment]
I'll add metrics and the option to configure the interval in a follow-up PR :+1:

File: zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java
Comment (lenaschoenburg) [reply zeebe/stream-platform/src/main/java/io/camunda/zeebe/stream/impl/ProcessingScheduleServiceImpl.java:213]: To be fair, IntelliJ warned me about this :D

**Pull request merged**
